src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder.cs

```diff
@@ -434,6 +434,46 @@ protected Task<ImmutableArray<Document>> FindDocumentsWithForEachStatementsAsync
             }
         }
 
+        protected async Task<ImmutableArray<ReferenceLocation>> FindReferencesInDeconstructionAsync(
+            ISymbol symbol,
+            Document document,
+            CancellationToken cancellationToken)
+        {
+            var syntaxTreeInfo = await SyntaxTreeIndex.GetIndexAsync(document, cancellationToken).ConfigureAwait(false);
+            if (!syntaxTreeInfo.ContainsDeconstruction)
+            {
+                return ImmutableArray<ReferenceLocation>.Empty;
+            }
+
+            var syntaxFacts = document.Project.LanguageServices.GetService<ISyntaxFactsService>();
+            var semanticFacts = document.Project.LanguageServices.GetService<ISemanticFactsService>();
+            var syntaxRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var locations = ArrayBuilder<ReferenceLocation>.GetInstance();
+
+            var originalUnreducedSymbolDefinition = symbol.GetOriginalUnreducedDefinition();
+
+            foreach (var node in syntaxRoot.DescendantNodesAndSelf())
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                var deconstructMethods = semanticFacts.GetDeconstructionAssignmentMethods(semanticModel, node);
+                if (deconstructMethods.IsEmpty)
+                {
+                    deconstructMethods = semanticFacts.GetDeconstructionForEachMethods(semanticModel, node);
+                }
+
+                if (deconstructMethods.Any(m => Matches(m, originalUnreducedSymbolDefinition)))
+                {
+                    var location = syntaxFacts.GetDeconstructionReferenceLocation(node);
+                    locations.Add(new ReferenceLocation(
+                        document, alias: null, location, isImplicit: true, isWrittenTo: false, CandidateReason.None));
```

Comment:
FYI, you should add a rename test (since rename uses the FAR engine).  because you used .isImplicit things should just work properly.  i.e. i think rename will complain about the implicit reference.  but i'm not sure.   But, at the very least, it will make sure we don't crash.  Tagging @jasonmalinowski @dpoeschl as they can give more insight here.

Reply:
That was useful. I did a manual test pass and got "Rename introduced errors in error-free code.".

For some reason, trying to renaming a `GetEnumerator` method doesn't pop this assert/dialog, despite the fact that it introduced an error in error-free code as well. I'll investigate.

Reply:
Yep, please add a test for this. :smile: